docs/API.rst

```diff
@@ -0,0 +1,83 @@
+API for ContestWebServer
+************************
+
+Authentication
+==============
+
+All the log-in methods and restrictions that apply to normal ContestWebServer
+usage also apply for API usage. In particular, this implies that IP autologin,
+if configured, can be used for API calls.
+
+Otherwise, making a `POST` request to the `/api/login` endpoint, passing
+`username` and `password` as form data (as is done for regular login), will
+return a JSON object with the following structure:
+
+.. sourcecode:: json
+
+  {"login_data": string}
+
+The content of the returned login data string should be passed to API calls
+that require authentication in a `X-CMS-Authorization` header.
+
+All other API methods described here require authentication.
+
+Task list
+=========
+
+An authenticated `GET` request to `/api/task_list`, done while the contest is
+running, will return the following object:
+
+.. sourcecode:: json
+
+  {"tasks": [{"name": string, "statements": [string], "submission_format": [string] }]}
+
+
+Tasks are ordered in the same order as in the UI. The `name` of the task is
+used for additional task-specific API calls. `statements` contains a list of
+languages for which a statement is available. `submission_format` contains a
+list of files that will need to be submitted, where filenames containing the
+`%l` string represent source files (that will be compiled and run on the
+server), and all other filenames represent output files.
+
+Submit
+======
+
+An authenticated `POST` request to `/api/{taskname}/submit` will send a
+submission. Files should be provided (according to the submission format) as
+form data, with field names matching the submission format.
+
+The request will return an object with the ID of the new submission:
+
+.. sourcecode:: json
+
+  {"id": number}
```

Comment:
We should make it clear what is the range of the numbers.

Ah, I just realized a nasty trap we are falling into: JSON is generally not transparent wrt. 64-bit numbers, because many libraries interpret them as floats. Let us make the submission IDs in the API strings, even though their contents will be numeric for the time being.

Reply:
Ugh, indeed. Fixed.